bundles/statistics/statsgrid/handler/DiagramHandler.js

```diff
@@ -0,0 +1,203 @@
+    constructor (stateHandler, service, sandbox) {
+        super();
+        this.stateHandler = stateHandler;
+        this.service = service;
+        this.sandbox = sandbox;
+        this.setState({
+            ...this.stateHandler.getState(),
+            sortOrder: 'value-descending',
+            diagramFlyout: null,
+            loading: false
+        });
+        this.loc = Oskari.getMsg.bind(null, 'StatsGrid');
+        this.addStateListener(() => this.updateFlyout());
+        this.eventHandlers = this.createEventHandlers();
+    };
+
+    getName () {
+        return 'DiagramHandler';
+    }
+
+    toggleDiagramFlyout (show, extraOnClose) {
+        if (show) {
+            if (!this.state.diagramFlyout) {
+                this.updateData();
+                this.showDiagramFlyout(extraOnClose);
+            }
+        } else {
+            this.closeDiagramFlyout();
+        }
+    }
+
+    showDiagramFlyout (extraOnClose) {
+        this.updateState({
+            diagramFlyout: showDiagramFlyout(this.getState(), this.getController(), () => {
+                this.closeDiagramFlyout();
+                if (extraOnClose) extraOnClose();
+            }),
+            sortOrder: 'value-descending'
+        });
+    }
+
+    closeDiagramFlyout () {
+        if (this.state.diagramFlyout) {
+            this.state.diagramFlyout.close();
+            this.updateState({
+                diagramFlyout: null,
+                sortOrder: 'value-descending'
+            });
+        }
+    }
+
+    updateFlyout () {
+        if (this.state.diagramFlyout) {
+            this.state.diagramFlyout.update(this.getState());
+        }
+    }
+
+    setSortOrder (order) {
+        this.updateState({
+            sortOrder: order
+        });
+    }
+
+    setActiveIndicator (hash) {
+        this.stateHandler.getController().setActiveIndicator(hash);
+    }
+
+    getColorScale (data) {
+        // Format data for Oskari.statistics.statsgrid.ClassificationService.getClassification
+        let numericData = {};
+        data.forEach((entry) => {
+            numericData[entry.id] = entry.value;
+        });
+        const activeIndicator = this.state.indicators?.find(indicator => indicator.hash === this.state.activeIndicator);
+        const { groups, bounds, error } = this.service.getClassificationService().getClassification(numericData, activeIndicator?.classification);
+        if (error) {
+            return ['#555'];
+        }
+        return {
+            bounds: bounds.slice(1, bounds.length - 1),
+            values: groups.map(group => group.color)
+        };
+    }
+
+    updateData () {
+        this.updateState({
+            loading: true
+        });
+        const indicator = this.service.getStateService().getActiveIndicator();
+
+        if (!indicator) return;
+        this.getIndicatorData(indicator, (data) => {
+            const isUndefined = (element) => {
+                return element.value === undefined;
+            };
```

Comment:
This should be moved out of the other function. Maybe just declare it at the start of the file?